src/mrd/ugbio_mrd/mrd_utils.py

```diff
@@ -708,3 +713,408 @@ def generate_synthetic_signatures(
+        if not np.isnan(x) and not np.isnan(y):
+            plt.text(
+                x,
+                y,
+                f"{y:.1e}" if y > zero_tf_fill else "0",
+                ha="right",
+                va="center",
+                color="b",
+                fontsize=16,
+            )  # draw above, centered
+    for line in hbp2["medians"]:
+        # get position data for median line
+        x, y = line.get_xydata()[0]  # top of median line
+        # overlay median value
+        if not np.isnan(x) and not np.isnan(y):
+            plt.text(
+                x,
+                y,
+                f"{y:.1e}" if y > zero_tf_fill else "0",
+                ha="right",
+                va="center",
+                color="g",
+                fontsize=16,
+            )  # draw above, centered
+
+
+def get_tf_from_filtered_data(
+    df_features_in: pd.DataFrame,
+    df_signatures_in: pd.DataFrame,
+    title=None,
+    denom_ratio=None,
+    *,
+    display_results=False,
+    plot_results=False,
+):
+    """
+    Calculate tumor fraction from filtered dataframes
+    Optionally display and plot the results
+    """
+    df_features_in_intersected = (
+        df_features_in.join(
+            df_signatures_in.groupby(level=["chrom", "pos"]).size().astype(bool).rename("locus_in_signature"),
+            how="inner",
+        )
+        .dropna(subset=["locus_in_signature"])
+        .query("locus_in_signature")
+        .drop(columns=["locus_in_signature"])
+    )  # retaun only loci that are in the signature df - they might have been filtered out
+    df_supporting_reads_per_locus = (
+        df_features_in_intersected.reset_index()
+        .groupby(["chrom", "pos", "signature", "signature_type"])
+        .size()
+        .rename("supporting_reads")
+        .reset_index(level=["signature", "signature_type"])
+    )
+    df_supporting_reads = (
+        (df_supporting_reads_per_locus.groupby(["signature_type", "signature"]).sum()).fillna(0).astype(int)
+    )
+    # fill in coverage for singatures with zero supporting reads
+    df_supporting_reads = (
+        pd.concat(
+            (
+                df_supporting_reads,
+                df_signatures_in.groupby(["signature_type", "signature"])["id"].sum().rename("supporting_reads"),
+            )
+        )
+        .fillna(0)
+        .astype(int)
+        .groupby(["signature_type", "signature"])
+        .sum()
+    )
+
+    df_coverage = (df_signatures_in.groupby("signature").agg({"coverage": "sum"})).fillna(0).astype(int)
+    df_tf = df_supporting_reads.join(df_coverage).fillna(0)
+    df_tf["corrected_coverage"] = df_tf["coverage"] * denom_ratio
+    df_tf["corrected_coverage"] = np.ceil(df_tf["corrected_coverage"])
+    df_tf = df_tf.assign(tf=df_tf["supporting_reads"] / df_tf["corrected_coverage"]).sort_index(ascending=False)
+
+    if plot_results:
+        plot_tf(df_tf, title=title)
+        plt.show()
+
+    if display_results:
```

Comment:
As above, the part that displays the table in a notebook should go in the notebook